akka-stream/src/main/scala/akka/stream/impl/package.scala

```diff
@@ -0,0 +1,461 @@
+ *    | | AddedModule | | AddedModule |  |
+ *
+ *  It is worth to note that the submodule might also continue this map-and-delegate chain to further submodules until
+ *  a builder is reached that can directly perform the operation. In other words, the depth of nesting is equal to
+ *  the length of mappedTo chaining.
+ *
+ *  IMPORTANT: When wiring in the enclosing module the new ports/shape MUST be used, using the original ports/shape
+ *  will lead to incorrect state.
+ *
+ *  == TraversalBuilders ==
+ *
+ *  In order to understand why builders are needed, consider wiring two ports together. Actually, we don't need to
+ *  wire input ports anywhere. Their slot is implicitly assigned by their position in the traversal, there is no
+ *  additional state we need to track. On the other hand, we cannot build a [[akka.stream.impl.MaterializeAtomic]]
+ *  node until the mapping array outToSlots is fully calculated. In other words, in reality, we don't wire input ports
+ *  anywhere, we only assign output ports to slots. The builders exist mainly to keep track all the necessary
+ *  information to be able to assign output ports, build the outToSlots array and finally the [[akka.stream.impl.MaterializeAtomic]]
+ *  node. The consequence of this that a [[akka.stream.impl.Traversal]] can be constructed as soon as all output ports
+ *  are wired ("unwired" inputs don't matter).
+ *
+ *  There is a specific builder that is used for the cases where all outputs have been wired:
+ *  [[akka.stream.impl.CompletedTraversalBuilder]]. This builder type simply contains the completed traversal plus
+ *  some additional data. The reason why this builder type exists is to keep auxiliary data structures required for
+ *  output port mapping only while they are needed, and shed them as soon as they are not needed anymore. Since builders
+ *  may recursively contain other builders, as soon as internals are completed those contained builders transition
+ *  to completed state and drop all additional data structures. This is very GC friendly as many intermediate graphs
+ *  exist only in a method call, and hence most of the additional data structures are dropped before method return and
+ *  can be efficiently collected by the GC.
+ *
+ *  The most generic builder is [[akka.stream.impl.CompositeTraversalBuilder]]. There are two main considerations
+ *  this builder needs to consider:
+ *
+ *   * Enclosed modules (builders) must have a stable position in the final traversal for relative addressing to work.
+ *     Since module offsets are calculated by traversal position, and outputs are wired relative to module offset, this
+ *     is critical.
+ *   * Enclosed builders might not be complete yet (i.e. have unwired outputs) and hence they cannot immediately give
+ *     a Traversal.
+ *
+ *  The composite builder keeps a temporary list of traversal steps (in reverse order because of immutable lists)
```

Comment:
add a note that while temporarily this builder keeps things in reverse, it will actually output a Traversal that contains modules in the same order they were added (unlike LinearTraversalBuilder)